### PR TITLE
Add support for sharedGalleryImageID in azure image reference

### LINKED ISSUE #71: Support "SharedGalleryImageId" field in Azure.
**What would you like to be added**:
To support “sharedGalleryImageId” field in Azure  
**Why is this needed**:

While “id” is already supported by Gardener, we still need Gardner to support “sharedGalleryImageId” field. 

Create a VM from a shared gallery image” in Virtual Machines :
https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fdocs.microsoft.com%2Fen-us%2Frest%2Fapi%2Fcompute%2Fvirtual-machines%2Fcreate-or-update%23create-a-vm-from-a-shared-gallery-image&data=05%7C01%7Coren.twaig%40sap.com%7C7d818b2a776b4be3f49808da5370f0a4%7C42f7676cf455423c82f6dc2d99791af7%7C0%7C0%7C637914041786174073%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=TDDln9puzp%2BLeiB5ZhwRNoCYybFrCp5ce07pVYL%2BIWo%3D&reserved=0

go SDK has support for the mentioned field but Gardener doesn't
https://github.com/Azure/azure-sdk-for-go/blob/37e48e769e089d25f401af65dc0f9960c12fd86d/services/compute/mgmt/2021-11-01/compute/models.go#L9665

This seems to have been introduced in v62.0.0:
https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAzure%2Fazure-sdk-for-go%2Fblob%2Fv62.0.0%2Fservices%2Fcompute%2Fmgmt%2F2021-11-01%2Fcompute%2Fmodels.go&data=05%7C01%7Coren.twaig%40sap.com%7C7d818b2a776b4be3f49808da5370f0a4%7C42f7676cf455423c82f6dc2d99791af7%7C0%7C0%7C637914041786174073%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=8u3Mr5vQZ%2FKpRzbvAnoKuotT5LReBqssQeWKGqxgh5o%3D&reserved=0

That format already tested by @I544024 , end it works with current Azure SDK in Gardner. (The path to the Shared Image: /SharedGalleries/82fc46df-cc38-4306-9880-504e872cee18-VSMP_MEMORYONE_GALLERY/Images/vSMP_MemoryONE/Versions/1062800000.0.0
According to Himanshu we should use PR request, to issue a request to support a new field in Gardner MCM and Azure extension (creation and validation). Adding the links:

https://github.com/gardener/gardener-extension-provider-azure/blob/master/pkg/apis/azure/types_worker.go#L50

https://github.com/gardener/gardener-extension-provider-azure/blob/d7c9f3ecf353dfd1e054e44e36d126df9c520da5/pkg/apis/azure/validation/cloudprofile.go#L66

https://github.com/gardener/machine-controller-manager-provider-azure/blob/4854a27c3c88201f9a1e755db749d9f88f4d4e8f/pkg/azure/apis/azure_provider_spec.go#L80

https://github.com/gardener/gardener-extension-provider-azure/pull/527

### PR DESCRIPTION
**Pull request by rishabh-11**
**What this PR does / why we need it**:

- Add support for machines based on shared gallery images

**Which issue(s) this PR fixes**:
Fixes #71 

**Special notes for your reviewer**:
SharedGalleryImageID eg:- `/SharedGalleries/<sharedGalleryName>/Images/<sharedGalleryImageName>/Versions/<sharedGalleryImageVersionName>`

**Release note**:

```improvement user
Machine-Controller-Manager Provider-Azure now supports managing virtual machines based on shared image gallery images. 
```

/invite @himanshu-kun

### CODE REVIEW

File: pkg/azure/apis/validation/validation.go [hunk 1]
```diff
@@ -192,13 +192,13 @@ func validateSecrets(secret *corev1.Secret) []error {
...
-		return append(allErrs, field.Required(fldPath, "must specify either a image id, community gallery image id or an urn"))
+	if isEmptyStringPtr(imageRef.URN) && isEmptyStringPtr(imageRef.CommunityGalleryImageID) && isEmptyString(imageRef.ID) && isEmptyStringPtr(imageRef.SharedGalleryImageID) {
+		return append(allErrs, field.Required(fldPath, "must specify either a image id, community gallery image id, shared gallery image id or an urn"))
 	}
 
 	if !isEmptyStringPtr(imageRef.URN) {
-		if !isEmptyStringPtr(imageRef.CommunityGalleryImageID) || !isEmptyString(imageRef.ID) {
-			return append(allErrs, field.Required(fldPath.Child("urn"), "cannot specify a urn and community gallery image id or image id in parallel"))
+		if !isEmptyStringPtr(imageRef.CommunityGalleryImageID) || !isEmptyString(imageRef.ID) || !isEmptyStringPtr(imageRef.SharedGalleryImageID) {
+			return append(allErrs, field.Required(fldPath.Child("urn"), "cannot specify a urn and community gallery image id, shared gallery image id or image id in parallel"))
```
Comment (himanshu-kun) [inline pkg/azure/apis/validation/validation.go:201]:
```suggestion
			return append(allErrs, field.Required(fldPath.Child("urn"), "cannot specify community gallery image id, shared gallery image id or image id when urn is specified"))
```

File: pkg/azure/apis/validation/validation.go [hunk 2]
```diff
@@ -216,13 +216,20 @@ func ValidateImageReference(imageRef api.AzureImageReference, fldPath *field.Pat
 	}
 
 	if !isEmptyStringPtr(imageRef.CommunityGalleryImageID) {
-		if !isEmptyString(imageRef.ID) {
-			return append(allErrs, field.Required(fldPath.Child("communityGalleryImageID"), "invalid community gallery image id format, empty field"))
+		if !isEmptyString(imageRef.ID) || !isEmptyStringPtr(imageRef.SharedGalleryImageID) {
+			return append(allErrs, field.Required(fldPath.Child("communityGalleryImageID"), "cannot specify a community image gallery id and shared image gallery id or image id in parallel"))
```
Comment (himanshu-kun) [inline pkg/azure/apis/validation/validation.go:220]:
```suggestion
			return append(allErrs, field.Required(fldPath.Child("communityGalleryImageID"), "cannot specify shared gallery image id or image id when community gallery image id is specified"))
```

**Review by himanshu-kun — CHANGES_REQUESTED** [review]

File: pkg/azure/apis/validation/validation.go (hunk 1 above)
Comment (rishabh-11) [reply pkg/azure/apis/validation/validation.go]: done

File: pkg/azure/apis/validation/validation.go (hunk 2 above)
Comment (rishabh-11) [reply pkg/azure/apis/validation/validation.go]: done

**Review by himanshu-kun — APPROVED** [review]
/lgtm

**Pull request merged**